### PR TITLE
Create `FOREIGN KEY` to `Identifier` for global links

### PR DESCRIPTION
**Pull request by nechaido**
<details>
<summary>Diff:</summary>

```diff
167,180d166
< -- Category: Catalog -----------------------------------------------------------
< 
< CREATE TABLE "Catalog" (
<   "Id"      bigint,
<   "Caption" text NOT NULL,
<   "Parent"  bigint
< );
< 
< CREATE INDEX "idxCatalogCaption" on "Catalog" ("Caption");
< 
< ALTER TABLE "Catalog" ADD CONSTRAINT "pkCatalogId" PRIMARY KEY ("Id");
< 
< ALTER TABLE "Catalog" ADD CONSTRAINT "fkCatalogParent" FOREIGN KEY ("Parent") REFERENCES "Catalog" ("Id") ON UPDATE RESTRICT ON DELETE SET NULL;
< 
196,197c182,192
< ALTER TABLE "Action" ADD CONSTRAINT "fkActionCategory" FOREIGN KEY ("Category") REFERENCES "Category" ("Id") ON UPDATE RESTRICT ON DELETE RESTRICT;
< ALTER TABLE "Category" ADD CONSTRAINT "fkCategoryMaster" FOREIGN KEY ("Master") REFERENCES "Category" ("Id") ON UPDATE RESTRICT ON DELETE SET NULL;
---
> -- Category: Catalog -----------------------------------------------------------
> 
> CREATE TABLE "Catalog" (
>   "Id"      bigint,
>   "Caption" text NOT NULL,
>   "Parent"  bigint
> );
> 
> CREATE INDEX "idxCatalogCaption" on "Catalog" ("Caption");
> 
> ALTER TABLE "Catalog" ADD CONSTRAINT "pkCatalogId" PRIMARY KEY ("Id");
266c261,268
< ALTER TABLE "Identifier" ADD CONSTRAINT "fkIdentifierCategory" FOREIGN KEY ("Category") REFERENCES "Category" ("Id") ON UPDATE RESTRICT ON DELETE SET NULL;
---
> 
> CREATE TABLE "CursorSubscribers" (
>   "Cursor"  bigint NOT NULL,
>   "Session" bigint NOT NULL
> );
> ALTER TABLE "CursorSubscribers" ADD CONSTRAINT "fkCursorSubscribersCursor" FOREIGN KEY ("Cursor") REFERENCES "Cursor" ("Id") ON UPDATE RESTRICT ON DELETE CASCADE;
> ALTER TABLE "CursorSubscribers" ADD CONSTRAINT "fkCursorSubscribersSession" FOREIGN KEY ("Session") REFERENCES "Identifier" ("Id") ON UPDATE RESTRICT ON DELETE CASCADE;
> ALTER TABLE "CursorSubscribers" ADD CONSTRAINT "pkSubscribersCursorSession" PRIMARY KEY ("Cursor", "Session");
296,297d297
< ALTER TABLE "Localization" ADD CONSTRAINT "fkLocalizationLanguage" FOREIGN KEY ("Language") REFERENCES "Language" ("Id") ON UPDATE RESTRICT ON DELETE RESTRICT;
< 
319a320,321
> ALTER TABLE "Locking" ADD CONSTRAINT "fkLockingSession" FOREIGN KEY ("Session") REFERENCES "Identifier" ("Id") ON UPDATE RESTRICT ON DELETE RESTRICT;
> ALTER TABLE "Locking" ADD CONSTRAINT "fkLockingSystemUser" FOREIGN KEY ("SystemUser") REFERENCES "Identifier" ("Id") ON UPDATE RESTRICT ON DELETE RESTRICT;
338,340d339
< ALTER TABLE "Permission" ADD CONSTRAINT "fkPermissionCategory" FOREIGN KEY ("Category") REFERENCES "Category" ("Id") ON UPDATE RESTRICT ON DELETE RESTRICT;
< ALTER TABLE "Permission" ADD CONSTRAINT "fkPermissionTarget" FOREIGN KEY ("Target") REFERENCES "Identifier" ("Id") ON UPDATE RESTRICT ON DELETE SET NULL;
< ALTER TABLE "Permission" ADD CONSTRAINT "fkPermissionCatalog" FOREIGN KEY ("Catalog") REFERENCES "Catalog" ("Id") ON UPDATE RESTRICT ON DELETE SET NULL;
346,347d344
< ALTER TABLE "PermissionActions" ADD CONSTRAINT "fkPermissionActionsPermission" FOREIGN KEY ("Permission") REFERENCES "Permission" ("Id") ON UPDATE RESTRICT ON DELETE CASCADE;
< ALTER TABLE "PermissionActions" ADD CONSTRAINT "fkPermissionActionsAction" FOREIGN KEY ("Action") REFERENCES "Action" ("Id") ON UPDATE RESTRICT ON DELETE CASCADE;
363,364d359
< ALTER TABLE "Replication" ADD CONSTRAINT "fkReplicationCategory" FOREIGN KEY ("Category") REFERENCES "Category" ("Id") ON UPDATE RESTRICT ON DELETE RESTRICT;
< 
377,378d371
< ALTER TABLE "Permission" ADD CONSTRAINT "fkPermissionRole" FOREIGN KEY ("Role") REFERENCES "Role" ("Id") ON UPDATE RESTRICT ON DELETE RESTRICT;
< 
395,396d387
< ALTER TABLE "Schema" ADD CONSTRAINT "fkSchemaCategory" FOREIGN KEY ("Category") REFERENCES "Category" ("Id") ON UPDATE RESTRICT ON DELETE RESTRICT;
< 
436c427
< ALTER TABLE "ServerNode" ADD CONSTRAINT "fkServerNodeServer" FOREIGN KEY ("Server") REFERENCES "Server" ("Id") ON UPDATE RESTRICT ON DELETE RESTRICT;
---
> ALTER TABLE "ServerNode" ADD CONSTRAINT "fkServerNodeServer" FOREIGN KEY ("Server") REFERENCES "Identifier" ("Id") ON UPDATE RESTRICT ON DELETE RESTRICT;
455,464d445
< ALTER TABLE "Locking" ADD CONSTRAINT "fkLockingSession" FOREIGN KEY ("Session") REFERENCES "Session" ("Id") ON UPDATE RESTRICT ON DELETE RESTRICT;
< 
< CREATE TABLE "CursorSubscribers" (
<   "Cursor"  bigint NOT NULL,
<   "Session" bigint NOT NULL
< );
< ALTER TABLE "CursorSubscribers" ADD CONSTRAINT "fkCursorSubscribersCursor" FOREIGN KEY ("Cursor") REFERENCES "Cursor" ("Id") ON UPDATE RESTRICT ON DELETE CASCADE;
< ALTER TABLE "CursorSubscribers" ADD CONSTRAINT "fkCursorSubscribersSession" FOREIGN KEY ("Session") REFERENCES "Session" ("Id") ON UPDATE RESTRICT ON DELETE CASCADE;
< ALTER TABLE "CursorSubscribers" ADD CONSTRAINT "pkSubscribersCursorSession" PRIMARY KEY ("Cursor", "Session");
< 
477,479d457
< ALTER TABLE "Permission" ADD CONSTRAINT "fkPermissionSubdivision" FOREIGN KEY ("Subdivision") REFERENCES "Subdivision" ("Id") ON UPDATE RESTRICT ON DELETE SET NULL;
< ALTER TABLE "Subdivision" ADD CONSTRAINT "fkSubdivisionParent" FOREIGN KEY ("Parent") REFERENCES "Subdivision" ("Id") ON UPDATE RESTRICT ON DELETE SET NULL;
< 
494,496d471
< ALTER TABLE "Replication" ADD CONSTRAINT "fkReplicationSource" FOREIGN KEY ("Source") REFERENCES "System" ("Id") ON UPDATE RESTRICT ON DELETE RESTRICT;
< ALTER TABLE "Replication" ADD CONSTRAINT "fkReplicationDestination" FOREIGN KEY ("Destination") REFERENCES "System" ("Id") ON UPDATE RESTRICT ON DELETE RESTRICT;
< 
511,513d485
< ALTER TABLE "Locking" ADD CONSTRAINT "fkLockingSystemUser" FOREIGN KEY ("SystemUser") REFERENCES "SystemUser" ("Id") ON UPDATE RESTRICT ON DELETE RESTRICT;
< ALTER TABLE "Session" ADD CONSTRAINT "fkSessionSystemUser" FOREIGN KEY ("SystemUser") REFERENCES "SystemUser" ("Id") ON UPDATE RESTRICT ON DELETE RESTRICT;
< ALTER TABLE "SystemUser" ADD CONSTRAINT "fkSystemUserSubdivision" FOREIGN KEY ("Subdivision") REFERENCES "Subdivision" ("Id") ON UPDATE RESTRICT ON DELETE SET NULL;
519,520d490
< ALTER TABLE "SystemUserRoles" ADD CONSTRAINT "fkSystemUserRolesSystemUser" FOREIGN KEY ("SystemUser") REFERENCES "SystemUser" ("Id") ON UPDATE RESTRICT ON DELETE CASCADE;
< ALTER TABLE "SystemUserRoles" ADD CONSTRAINT "fkSystemUserRolesRole" FOREIGN KEY ("Role") REFERENCES "Role" ("Id") ON UPDATE RESTRICT ON DELETE CASCADE;

```

</details>

### CODE REVIEW

File: lib/pg.ddl.js [hunk 1]
```diff
@@ -222,15 +236,15 @@ const generateProperties =
 //   type - string, type of link
 //   link - object
 // Returns: string
-const generateLink = (from, to, name, type, link) => {
+const generateLink = (from, to, name, type, link, destination) => {
```
Comment (belochub) [inline lib/pg.ddl.js]: You should also update the comments above.

File: lib/pg.ddl.js [hunk 2]
```diff
@@ -252,7 +266,9 @@ const generateLink = (from, to, name, type, link) => {
 //   link - object
 //   categories - object, object that contains categories by their name
 // Returns: string
-const generateManyToMany = (from, to, name, link, categories) => {
+const generateManyToMany = (
+  from, to, name, link, categories, destination, source
```
Comment (belochub) [inline lib/pg.ddl.js]: Ditto.

File: lib/pg.ddl.js [hunk 3]
```diff
@@ -146,10 +142,13 @@ const categorizeEntries = (category, categoryName, categories) => {
...
           categories,
           property,
-          categoryName,
-          property.category
+          categoryName
         );
 
+        entry.destination = isGlobalCategory(property.definition) ?
+          'Identifier' :
+          property.category;
```
Comment (tshemsedinov) [inline lib/pg.ddl.js]: to single line

File: lib/pg.ddl.js [hunk 4]
```diff
@@ -212,16 +211,17 @@ const generateProperties =
 //   name - string, name of a link
 //   type - string, type of link
 //   link - object
+//   destination - string, name of a table to crate `FOREIGN KEY` to
 // Returns: string
-const generateLink = (from, to, name, type, link) => {
+const generateLink = (from, to, name, type, link, destination) => {
```
Comment (tshemsedinov) [inline lib/pg.ddl.js]: remove link

File: lib/pg.ddl.js [hunk 5]
```diff
@@ -242,8 +242,9 @@ const generateLink = (from, to, name, type, link) => {
 //   type - string, type of link
 //   link - object
 //   categories - object, object that contains categories by their name
+//   destination - string, name of a table to crate `FOREIGN KEY` to
 // Returns: string
-const generateManyToMany = (from, to, name, link, categories) => {
+const generateManyToMany = (from, to, name, link, categories, destination) => {
```
Comment (tshemsedinov) [inline lib/pg.ddl.js]: ditto

File: lib/pg.ddl.js [hunk 6]
```diff
@@ -212,20 +211,21 @@ const generateProperties =
 //   name - string, name of a link
 //   type - string, type of link
 //   link - object
```
Comment (belochub) [inline lib/pg.ddl.js]: Ditto.

File: lib/pg.ddl.js [hunk 7]
```diff
@@ -212,20 +211,21 @@ const generateProperties =
...
+//   destination - string, name of a table to crate `FOREIGN KEY` to
```
Comment (belochub) [inline lib/pg.ddl.js]:
```suggestion
//   destination - string, name of a table to create `FOREIGN KEY` to
```

File: lib/pg.ddl.js [hunk 8]
```diff
@@ -242,8 +242,9 @@ const generateLink = (from, to, name, type, link) => {
 //   type - string, type of link
 //   link - object
```
Comment (belochub) [inline lib/pg.ddl.js]: Ditto.

File: lib/pg.ddl.js [hunk 9]
```diff
@@ -242,8 +242,9 @@ const generateLink = (from, to, name, type, link) => {
 //   type - string, type of link
```
Comment (belochub) [inline lib/pg.ddl.js]: Ditto.

**Comment by nechaido** [comment]
@belochub @tshemsedinov

**Comment by belochub** [comment]
@nechaido can you rebase on `master`, please?

**Comment by nechaido** [comment]
@belochub done.

File: lib/pg.ddl.js [hunk 10]
```diff
@@ -486,21 +484,42 @@ const generateTables = (categories, types) => {
...
   const history = Object.assign({}, schema);
-  history._Creation = { domain: 'DateTime', required: true, index: true };
-  history._Effective = { domain: 'DateTime', required: true, index: true };
-  history._Cancel = { domain: 'DateTime', index: true };
-  history._HistoryStatus = { domain: 'HistoryStatus', required: true };
+  history._Creation = {
+    domain: 'DateTime',
+    required: true,
+    index: true,
+    definition: domains.get('DateTime'),
```
Comment (belochub) [inline lib/pg.ddl.js]: I think you can save the value of `domains.get('DateTime')` into the variable to avoid calling `get()` multiple times here.

**Comment by nechaido** [comment]
ping @tshemsedinov.

**Pull request merged**